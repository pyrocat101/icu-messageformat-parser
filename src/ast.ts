export const enum TYPE {
  /**
   * Raw text
   */
  literal,
  /**
   * Variable w/o any format, e.g `var` in `this is a {var}`
   */
  argument,
  /**
   * Variable w/ number format
   */
  number,
  /**
   * Variable w/ date format
   */
  date,
  /**
   * Variable w/ time format
   */
  time,
  /**
   * Variable w/ select format
   */
  select,
  /**
   * Variable w/ plural format
   */
  plural,
}

export interface LocationDetails {
  offset: number;
  line: number;
  column: number;
}
export interface Location {
  start: LocationDetails;
  end: LocationDetails;
}

export interface BaseElement<T extends TYPE> {
  type: T;
  value: string;
  location?: Location;
}

export type LiteralElement = BaseElement<TYPE.literal>;
export type ArgumentElement = BaseElement<TYPE.argument>;

export interface SimpleFormatElement<T extends TYPE> extends BaseElement<T> {
  style?: string;
}

export type NumberElement = SimpleFormatElement<TYPE.number>;
export type DateElement = SimpleFormatElement<TYPE.date>;
export type TimeElement = SimpleFormatElement<TYPE.time>;

export interface SelectOption {
  id: string;
  value: MessageFormatElement[];
  location?: Location;
}

export type ValidPluralRule = 'zero' | 'one' | 'two' | 'few' | 'many' | 'other' | string;

export interface PluralOrSelectOption {
  value: MessageFormatElement[];
  location?: Location;
}

export interface SelectElement extends BaseElement<TYPE.select> {
  options: Record<string, PluralOrSelectOption>;
}

export interface PluralElement extends BaseElement<TYPE.plural> {
  options: Record<ValidPluralRule, PluralOrSelectOption>;
  offset: number;
  pluralType: Intl.PluralRulesOptions['type'];
}

export type MessageFormatElement =
  | LiteralElement
  | ArgumentElement
  | NumberElement
  | DateElement
  | TimeElement
  | SelectElement
  | PluralElement;
