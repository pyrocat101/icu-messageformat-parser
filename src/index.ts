export * from './types';
export * from './parser';
export * from './printer';
