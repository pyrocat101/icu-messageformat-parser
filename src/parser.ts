import {MessageFormatElement, LiteralElement, TYPE} from './ast';

/**
 * MessageFormat EBNF:
 *
 * message = messageText (argument messageText)*
 * argument = noneArg | simpleArg | complexArg
 * complexArg = choiceArg | pluralArg | selectArg | selectordinalArg
 *
 * noneArg = '{' argNameOrNumber '}'
 * simpleArg = '{' argNameOrNumber ',' argType [',' argStyle] '}'
 * choiceArg = '{' argNameOrNumber ',' "choice" ',' choiceStyle '}'
 * pluralArg = '{' argNameOrNumber ',' "plural" ',' pluralStyle '}'
 * selectArg = '{' argNameOrNumber ',' "select" ',' selectStyle '}'
 * selectordinalArg = '{' argNameOrNumber ',' "selectordinal" ',' pluralStyle '}'
 *
 * choiceStyle: see ChoiceFormat
 * pluralStyle: see PluralFormat
 * selectStyle: see SelectFormat
 *
 * argNameOrNumber = argName | argNumber
 * argName = [^[[:Pattern_Syntax:][:Pattern_White_Space:]]]+
 * argNumber = '0' | ('1'..'9' ('0'..'9')*)
 *
 * argType = "number" | "date" | "time" | "spellout" | "ordinal" | "duration"
 * argStyle = "short" | "medium" | "long" | "full" | "integer" | "currency" | "percent" | argStyleText | "::" argSkeletonText
 *
 */

interface InputState {
  readonly text: string;
  readonly pos: number;
}

// TODO: parse error
type ParserMonad<TResult> = {
  readonly input: InputState;
  readonly result: TResult;
} | null;
type Parser<TResult> = (input: InputState) => ParserMonad<TResult>;

function consumeRegex(input: InputState, pattern: RegExp): ParserMonad<RegExpExecArray> {
  const {text, pos} = input;
  pattern.lastIndex = pos;
  const match = pattern.exec(text);
  return match && {input: {text, pos: pos + match[0].length}, result: match};
}

function consumeText(input: InputState, prefix: string): ParserMonad<string | null> {
  const {text, pos} = input;
  return text.startsWith(prefix, pos)
    ? {input: {text, pos: pos + prefix.length}, result: prefix}
    : null;
}

export const parseMessage: Parser<MessageFormatElement[]> = input => {
  const chunks: MessageFormatElement[] = [];
  while (true) {
    const parsed = parseMessageText(input);
    if (parsed != null) {
      input = parsed.input;
      chunks.push(parsed.result);
    } else {
      break;
    }
  }
  return {input, result: chunks};
};

const ESCAPED_TEXT = /'\{(?:[^']|(?:''))+'/suy;
const UNESCAPED_MSG_TEXT = /(?:(?!''|\{).)+/suy;

// TODO: use macro.
// ICU >= 4.8 quoting behavior.
const parseMessageText: Parser<LiteralElement> = input => {
  const chunks: string[] = [];
  while (true) {
    // Starting with ICU 4.8, an ASCII apostrophe only starts quoted text if it immediately precedes
    // a character that requires quoting (that is, "only where needed"), and works the same in
    // nested messages as on the top level of the pattern. The new behavior is otherwise compatible.
    {
      // TODO: other syntax characters?
      const parsed = consumeRegex(input, ESCAPED_TEXT);
      if (parsed != null) {
        input = parsed.input;
        chunks.push(parsed.result[0].slice(1, -1).replace("''", "'"));
        continue;
      }
    }
    // double apostrophes
    {
      const parsed = consumeText(input, "''");
      if (parsed != null) {
        input = parsed.input;
        chunks.push("'");
        continue;
      }
    }
    // unescaped string
    {
      const parsed = consumeRegex(input, UNESCAPED_MSG_TEXT);
      if (parsed != null) {
        input = parsed.input;
        chunks.push(parsed.result[0]);
        continue;
      }
    }
    break;
  }
  // We need to actually consume input.
  return chunks.length > 0
    ? {
        input,
        result: {
          type: TYPE.literal,
          value: chunks.join(''),
          // TODO: location
        },
      }
    : null;
};
