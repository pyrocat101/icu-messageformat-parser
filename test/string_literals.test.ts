import {parseMessage} from '../src/parser';

test('unescaped string literals', () => {
  expect(parseMessage({text: 'line1\n  test2', pos: 0})).toMatchSnapshot();
});

test('treats double apostrophes as one apostrophe', () => {
  expect(parseMessage({text: `a''b`, pos: 0})).toMatchSnapshot();
});

test('starts quoted text if apostrophe immediately precedes a character', () => {
  expect(parseMessage({text: `'{a''b}'`, pos: 0})).toMatchSnapshot();
});

test('does not start quoted text if apostrophe does not immediately precede a character', () => {
  expect(parseMessage({text: `'a}a''b}'`, pos: 0})).toMatchSnapshot();
});
