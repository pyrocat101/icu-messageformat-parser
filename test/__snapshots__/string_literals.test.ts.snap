// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`does not start quoted text if apostrophe does not immediately precede a character 1`] = `
Object {
  "input": Object {
    "pos": 9,
    "text": "'a}a''b}'",
  },
  "result": Array [
    Object {
      "type": 0,
      "value": "'a}a'b}'",
    },
  ],
}
`;

exports[`starts quoted text if apostrophe immediately precedes a character 1`] = `
Object {
  "input": Object {
    "pos": 8,
    "text": "'{a''b}'",
  },
  "result": Array [
    Object {
      "type": 0,
      "value": "{a'b}",
    },
  ],
}
`;

exports[`treats double apostrophes as one apostrophe 1`] = `
Object {
  "input": Object {
    "pos": 4,
    "text": "a''b",
  },
  "result": Array [
    Object {
      "type": 0,
      "value": "a'b",
    },
  ],
}
`;

exports[`unescaped string literals 1`] = `
Object {
  "input": Object {
    "pos": 13,
    "text": "line1
  test2",
  },
  "result": Array [
    Object {
      "type": 0,
      "value": "line1
  test2",
    },
  ],
}
`;
